src/ck-core/ckfutures.C

```diff
@@ -20,20 +20,100 @@ in remote process control.
 #include "ckfutures.h"
 #include <stdlib.h>
 #include <limits>
+#include <memory>
 
-typedef struct Future_s {
-  bool ready;
-  void *value;
-  CthThread waiters;
-  int next; 
-} Future;
-
-typedef struct {
-  Future *array;
-  int max;
-  int freelist;
-}
-FutureState;
+struct FutureRequest {
+  virtual void fulfill(void* value) = 0;
+};
+
+class FutureToThread: public FutureRequest {
+  CthThread th;
+ public:
+  FutureToThread(CthThread th_) : th(th_) {}
+
+  virtual void fulfill(void* value) override {
+    if (th) CthAwaken(th);
+    th = nullptr;
+  }
+};
+
+class FutureToFuture: public FutureRequest {
+  CkFuture fut; 
+  bool fulfilled;
+ public:
+  FutureToFuture(CkFuture fut_) : fut(fut_), fulfilled(false) {}
+
+  virtual void fulfill(void* value) override {
+    if (!fulfilled) CkSendToFuture(fut, value);
+    fulfilled = true;
+  }
+};
+
+struct FutureState {
+  using request_t = std::shared_ptr<FutureRequest>;
+  using request_queue_t = std::vector<request_t>;
+ private:
+  std::unordered_map<CkFutureID, void*> values;
+  std::unordered_map<CkFutureID, request_queue_t> waiting;
+  std::set<CkFutureID> freeList;
```

Comment:
I might be more efficient to use a `std::vector`, `std::queue` or `std::stack` here. Do you need the semantics of a set for something?

Reply:
A set was chosen to preclude issues with repeated frees of the same future. If this is something we would rather specify as undefined behavior, a queue (or any of the others could be chosen).

Reply:
I see. I think it's fine to disallow repeated frees (the old semantics didn't allow it), but I'll leave the choice up to you.

Reply:
I am leaning towards changing it to a vector and adding `CkAssert(!(freeList contains id) && "repeated frees of the same future");` so if error-checking is enabled it'll notify the user about repeated frees and, otherwise, will not have have any performance impact.